### ISSUE TITLE
TN3270 fails to connect to z/OS host

### ISSUE THREAD

**Issue by John-A-Davies** [zowe/zlux#592]
**Describe the bug**
There's something wrong with the way TN3270 starts and connects. When I start the TN3270 window, it opens but doesn't connect to the host. But if I toggle the on/off switch, it connects right away.

Here's the STDOUT at that time:
```
11:18:01.250 <ZWED:66038> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET                  
11:19:08.404 <ZWED:66039> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET                  
11:19:08.421 <ZWED:66039> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0101I - [Host=x.x.x.x, Port=23, ClientIP=x.x.x.x
11:19:08.425 <ZWED:66039> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0102I - Total TN3270 terminals connected: 1     
```           

it's the attempt at 11:18:01.250 that doesn't connect.

However, sometimes toggling the ON/OFF still doesn't make it connect, and sometimes it connects right away.

**To Reproduce**
1. Start MVD
2. Log in
3. Open TN3270 window
4. TN3270 window opens
5. Blinking cursor appears, but screen is otherwise blank
6. Reveal the configuration bar at the top of the TN3270 screen
7. Click the green ON/OFF button to disconnect
8. Click the grey   ON/OFF button to connect
9. TN3270 connects to host, and you can see the z/OS screen where you can log in

**Expected behavior**
TN3270 connects in under 5 seconds.

**Screenshots**

![image](https://user-images.githubusercontent.com/36740892/106118733-5700a180-614c-11eb-9ec3-f21aa861ae93.png)

**Additional context**
My theory is that the connect handshake becomes deadlocked due to a race condition, and ON/OFF frees the deadlock.  It's not due to slow data transfer between the node app and the TCP port on z/OS, even though this is a ZD&T system.

**Comment by John-A-Davies** [comment]
This showed up in v1.19 RC 2 but I don't think it's a regression - I've seen this kind of delay to connect in TN3270 and VT before but I never got to the bottom of it.  I didn't find any related Git issues.

**Comment by 1000TurquoisePogs** [comment]
With the increased logging (`ZWED_logLevels_org____zowe____terminal____proxy_____x2a=5` in instance.env and `ZoweZLUX.logger.setLogLevelForComponentName("org.zowe.terminal.core.telnet", 5)` and `ZoweZLUX.logger.setLogLevelForComponentName("org.zowe.terminal.core.protocol", 5)` in the browser) did you see anything more about where things just stopped? If the browser never got anything, then it's stuck server-side and that would mean it's just a TCP socket issue not specific to our code. Best we could do is just do a retry loop until a timeout.

**Comment by John-A-Davies** [comment]
I enabled the web console to view the network requests.  Normally I get 200 and TN3270 connects.

I tried opening tn3270 a few times until it got stuck, whereupon I got this in the network:
```
GET
	wss://9.20.205.110:8544/ZLUX/plugins/org.zowe.terminal.tn3270/services/terminalstream/_current/
Status    101
Switching Protocols
Version    HTTP/1.1
Transferred129 B (0 B size)
```
Each time I toggle ON/OFF, I get that 101.  Eventually I get 200 and it connects.

**Comment by 1000TurquoisePogs** [comment]
HTTP 101 must happen every single time if successful.
HTTP 101 is establishment of the websocket, which is essential to the communication of the terminal
Websockets are like TCP sockets but embedded within HTTP
The network traffic will appear to be

Client: HTTP GET with ws:// or wss:// at front, which has the "Upgrade" header
Server: responds with HTTP 101, switching protocols... from then onward, connection is websocket instead of HTTP

**Comment by John-A-Davies** [comment]
Yes, a 200 is always preceded by a 101.  But sometimes it hangs on 101, i.e. the screen does not get populated from the host.  Then if you click ON/OFF, another 101 happens, followed by 200 and the screen gets filled.  How many tries you need to connect appears random; it only hangs sometimes, maybe 1 out of 10.

**Comment by 1000TurquoisePogs** [comment]
Which url is the 200 associated with? If it's the tn3270 discovery service on ZSS, then while it does indicate a successful connection, it doesn't tell you why something would fail if it did. It's an independent service that gets triggered as the terminal session is ongoing as a way to sense the panel you're on for potential cool future integrations.

**Comment by 1000TurquoisePogs** [comment]
The conversation flow is:
1) client opens websocket to server
2) server accepts it and awaits commands
3) client sends a "connect" command, saying which destination the server should mediate https://github.com/zowe/zlux-server-framework/blob/staging/plugins/terminal-proxy/lib/terminalProxy.js#L286
4) a TCP socket is established to the destination https://github.com/zowe/zlux-server-framework/blob/72113b8775f9dbe802c191443eeadfa76f4e4dd0/plugins/terminal-proxy/lib/terminalProxy.js#L623
5) (optionally) TLS startup actions associated with the new socket are also done  https://github.com/zowe/zlux-server-framework/blob/72113b8775f9dbe802c191443eeadfa76f4e4dd0/plugins/terminal-proxy/lib/terminalProxy.js#L514-L569

6) on socket success, the app server becomes a simple proxy by which it sends whatever comes out of the tcp socket to the browser via websocket, and the browser reacts to it or sends fresh user input back through the websocket, that the server forwards through the tcp socket

**Comment by 1000TurquoisePogs** [comment]
When 6 happens, this is where the previously mentioned logging comes in. You'll start to see a lot of bytes. If not, then we have a socket issue. Something between nodejs and the tn3270 sever are in disagreement?

**Comment by John-A-Davies** [comment]
I captured the browser console log.  

The order of actions was
1) open the TN3270 terminal ( it did not connect)
2) set the logs
3) press the power button, keep retrying until you trigger the issue.

 
[defect1926.console.log.txt](https://github.com/zowe/zlux/files/5893158/defect1926.console.log.txt)
In that log, 10:10:40.229 is the first time the ON/OFF toggle produced a successful connection; prior attempts at ON/OFF showed a blank window.

**Comment by John-A-Davies** [comment]
The corresponding STDOUT from the zowe server JES log is below

```
2021-01-29 09:42:53.565 <ZWED:66537> ZWESVUSR INFO (_zsf.auth,native array.js:958) ZWED0070I - uN0yKXuRCRpAc88XFddUPmkFpGWIRY22: Session security call authentic
2021-01-29 09:42:53.641 <ZWED:66537> ZWESVUSR INFO (_zsf.auth,native array.js:958) ZWED0070I - uN0yKXuRCRpAc88XFddUPmkFpGWIRY22: Session security call authentic
2021-01-29 09:42:54.796 <ZWED:66537> ZWESVUSR INFO (org.zowe.zlux.bootstrap:adminnotificationdata,native array.js:958) ZWED0001IGET                             
2021-01-29 09:42:55.267 <ZWED:66537> ZWESVUSR INFO (_zsf.auth,native array.js:958) ZWED0070I - uN0yKXuRCRpAc88XFddUPmkFpGWIRY22: Session security call refreshSt
2021-01-29 09:42:55.281 <ZWED:66537> ZWESVUSR INFO (_zsf.auth,native array.js:958) ZWED0070I - uN0yKXuRCRpAc88XFddUPmkFpGWIRY22: Session security call refreshSt
2021-01-29 09:43:47.670 <ZWED:66537> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:10:03.480 <ZWED:66538> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:10:22.379 <ZWED:66537> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:10:36.831 <ZWED:66538> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:10:50.638 <ZWED:66537> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:11:03.450 <ZWED:66538> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0106I - Saw Websocket request, method=GET       
2021-01-29 10:11:03.474 <ZWED:66538> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0101I - [Host=9.20.205.110, Port=23, ClientIP=9.
2021-01-29 10:11:03.479 <ZWED:66538> ZWESVUSR INFO (org.zowe.terminal.proxy:tn3270data,native array.js:958) ZWED0102I - Total TN3270 terminals connected: 1     
```
